### PR TITLE
Refactor storage unit

### PR DESCRIPTION
**Pull request by ssd04**
Separate factory package for `db` `cache`.
Refactor and cleanup storageunit package.

### CODE REVIEW

File: factory/cache.go [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+			return nil, fmt.Errorf("%w, provided %d, minimum %d",
+				common.ErrLRUCacheInvalidSize,
+				sizeInBytes,
+				minimumSizeForLRUCache,
+			)
+		}
+
+		cacher, err = lrucache.NewCacheWithSizeInBytes(int(capacity), int64(sizeInBytes))
+	case common.FIFOShardedCache:
+		cacher, err = fifocache.NewShardedCache(int(capacity), int(shards))
```
Comment (iulianpascalau) [inline factory/cache.go]: return directly?

File: factory/cache.go [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+	case common.SizeLRUCache:
+		if sizeInBytes < minimumSizeForLRUCache {
+			return nil, fmt.Errorf("%w, provided %d, minimum %d",
+				common.ErrLRUCacheInvalidSize,
+				sizeInBytes,
+				minimumSizeForLRUCache,
+			)
+		}
+
+		cacher, err = lrucache.NewCacheWithSizeInBytes(int(capacity), int64(sizeInBytes))
```
Comment (iulianpascalau) [inline factory/cache.go]: return directly?

File: factory/cache.go [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+	case common.FIFOShardedCache:
+		cacher, err = fifocache.NewShardedCache(int(capacity), int(shards))
+		if err != nil {
+			return nil, err
+		}
+	default:
+		return nil, common.ErrNotSupportedCacheType
+	}
+
+	if err != nil {
```
Comment (iulianpascalau) [inline factory/cache.go]: L53-L57 not required if we will return directly in the `case` instructions

File: factory/cache_test.go [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+	assert.NotNil(t, err, "error expected")
+	assert.Nil(t, cacher, "cacher expected to be nil, but got %s", cacher)
+}
+
+func TestCreateCacheFromConfOK(t *testing.T) {
+
+	cacher, err := factory.NewCache(common.CacheConfig{Type: common.LRUCache, Capacity: 10, Shards: 1, SizeInBytes: 0})
+
+	assert.Nil(t, err, "no error expected but got %s", err)
+	assert.NotNil(t, cacher, "valid cacher expected but got nil")
```
Comment (iulianpascalau) [inline factory/cache_test.go]: Please add more tests here actually testing the result cachers are of the expected type. Can use `fmt.Sprintf("%T", cacher)` to consistently check the type.

File: factory/db_test.go [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+		BatchDelaySeconds: 10,
+		MaxBatchSize:      10,
+		MaxOpenFiles:      10,
+	}
+	persister, err := factory.NewDB(argsDB)
+	assert.Nil(t, err, "no error expected")
+	assert.NotNil(t, persister, "valid persister expected but got nil")
+
+	err = persister.Destroy()
+	assert.Nil(t, err, "no error expected destroying the persister")
```
Comment (iulianpascalau) [inline factory/db_test.go]: same as with the cachers, please add more unit tests

File: factory/cache.go (hunk 3 above)
Comment (ssd04) [reply factory/cache.go]: updated the entire function

File: factory/cache_test.go (hunk 4 above)
Comment (ssd04) [reply factory/cache_test.go]: added more unit tests

File: factory/db_test.go (hunk 5 above)
Comment (ssd04) [reply factory/db_test.go]: added more unit tests here also

File: factory/cache_test.go [hunk 6]
```diff
@@ -1,25 +1,97 @@
...
+
+		cacheConf := common.CacheConfig{
+			Type:        common.LRUCache,
+			Capacity:    100,
+			Shards:      1,
+			SizeInBytes: 0,
+		}
+		cacher, err := factory.NewCache(cacheConf)
+		require.Nil(t, err)
+		require.Equal(t, "*lrucache.lruCache", fmt.Sprintf("%T", cacher))
```
Comment (iulianpascalau) [inline factory/cache_test.go:53]: :+1:

File: factory/db_test.go [hunk 7]
```diff
@@ -1,63 +1,98 @@
...
+			BatchDelaySeconds: 10,
+			MaxBatchSize:      10,
+			MaxOpenFiles:      10,
+		}
+		persister, err := factory.NewDB(argsDB)
+		assert.NotNil(t, err)
+		assert.Nil(t, persister)
+	})
+
+	t.Run("LvlDB type, should work", func(t *testing.T) {
```
Comment (iulianpascalau) [inline factory/db_test.go:48]: I guess for this tests a persister.Close should be called to free the resources

**Review by iulianpascalau — DISMISSED** [review]

File: factory/db_test.go (hunk 7 above)
Comment (ssd04) [reply factory/db_test.go:48]: added persister close

**Review by sstanculeanu — APPROVED** [review]

**Review by iulianpascalau — APPROVED** [review]

**Pull request merged**